### PR TITLE
CONSOLE-3171, CONSOLE-3329: Re-initialize graphql client on cluster change

### PR DESCRIPTION
**Pull request by rawagner**

### CODE REVIEW

File: frontend/packages/console-app/src/components/nav/NavHeader.tsx
Comment (rawagner) [inline frontend/packages/console-app/src/components/nav/NavHeader.tsx:125]: I think we should create a function which will do all the necessary dispatches as these 3 function calls are repeating in couple more places.

Also I think we should re-run API discovery

File: pkg/server/server.go
Comment (rawagner) [inline pkg/server/server.go:365]: @TheRealJon you already had the backend part implemented :)

File: frontend/public/graphql/client.ts
Comment (TheRealJon) [inline frontend/public/graphql/client.ts:26]: Could we pass the cluster in as an arg? Technically, we are only using session storage to persist the active cluster across page reloads. If we ever changed that, it would break this client.

File: frontend/public/graphql/client.ts
Comment (TheRealJon) [inline frontend/public/graphql/client.ts:57]: nit, descriptive names:
```suggestion
    fetch: (url: string, options) => {
```

File: pkg/server/server.go
Comment (TheRealJon) [reply pkg/server/server.go:365]: I'm about to change this implementation a little, but it should be an easy adjustment. Instead of creating a new proxy instance on every request, we should create one on the first request, then reuse the same proxy instance on subsequent requests.

File: frontend/packages/console-app/src/components/nav/NavHeader.tsx
Comment (TheRealJon) [reply frontend/packages/console-app/src/components/nav/NavHeader.tsx:125]: Feel free to add the api discovery here. We only left it out because we knew it wouldn't work yet :)

**Review by TheRealJon — COMMENTED** [review]
A couple of questions and nits, but this is awesome!

File: pkg/server/server.go
Comment (rawagner) [reply pkg/server/server.go:365]: that is happening already IIUC. The handle function is executed once, when graphl client tries to connect to graphql server - we create a proxy and graphql handler here. The rest of the communication is done via websocket and the graphql handler is always using the same proxy instance.
When the cluster is changed in the UI, the websocket is closed and new one is created, executing the handle function again (with new cluster as parameter)

File: frontend/public/graphql/client.ts
Comment (rawagner) [reply frontend/public/graphql/client.ts:57]: updated

File: frontend/packages/console-app/src/components/nav/NavHeader.tsx
Comment (rawagner) [reply frontend/packages/console-app/src/components/nav/NavHeader.tsx:125]: I will create a separate PR for that. There is some caching involved on frontend side so Im not sure that running just `dispatch(startAPIDiscovery())` will be correct.

File: frontend/public/graphql/client.ts
Comment (rawagner) [reply frontend/public/graphql/client.ts:26]: yes, i agree. That will be much better

**Review by jhadvig — APPROVED** [review]
/lgtm

QE Approver:
 /assign @yapei
Docs Approver:
 /assign @opayne1
PX Approver:
/assign @RickJWagner

**Review by opayne1 — COMMENTED** [review]
/label docs-approved

File: pkg/server/server.go
Comment (TheRealJon) [reply pkg/server/server.go:365]: Yes, but I think the number of proxy instances still scales up with the number of requests, which is probably not a scalable solution. With many users/sessions, proxy instances increase.

**Comment by yapei** [comment]
@rawagner Could you help rebase to master?

**Comment by yapei** [comment]
websocket connection to `/api/graphql` is re-established whenever we switch to managed cluster
/label qe-approved

**Review by TheRealJon — COMMENTED** [review]
During our testathon, these changes were deployed and a few regressions were noted that are related:

- User settings fail to load when managed clusters are selected since the user UID changes. We are only fetching the user settings configmap from the hub cluster, and we use the user UID as part of the configmap name. We probably need to either track the hub cluster user UID separately to make sure we fetch the correct configmap, or update the user settings logic to fall back to local storage in a multicluster environment.
- I experienced 401 responses from graphql when logged in as a user with cluster viewer permissions.
- API discovery still needs to be addressed.

We can probably merge this and address the issues in a follow-on PR, but I just wanted to make sure we have these documented.

**Comment by jhadvig** [comment]
/hold
until we figure out what needs to be done with issues, mentioned in [Jon's](https://github.com/openshift/console/pull/12425#pullrequestreview-1299979300) comment

**Comment by rawagner** [comment]
Great findings, @TheRealJon 
1 and 3 are pretty clear but Im not sure why 2 would occur. I dont know anything about cluster-viewer role.
Can we create sub-tasks for all of these and address them separately ?

**Comment by TheRealJon** [comment]
> Great findings, @TheRealJon 1 and 3 are pretty clear but Im not sure why 2 would occur. I dont know anything about cluster-viewer role. Can we create sub-tasks for all of these and address them separately ?

I'm hesitant to merge this without at least addressing 1 and 2 because those are pretty serious regressions. The API discovery issue might already have a bug in Jira and I'm fine tracking it separately since it's not introduced by these changes.

**Comment by TheRealJon** [comment]
Multicluster feature has been deferred. Closing.

**Comment by TheRealJon** [comment]
Both related Jira issues:
https://issues.redhat.com//browse/CONSOLE-3329
https://issues.redhat.com//browse/CONSOLE-3171

**Pull request closed without merging**
